### PR TITLE
Stencil 3250 - Webpack 2 w/ Code Splitting, Tree Shaking

### PR DESCRIPTION
**Pull request by mjschock**
### What?

Webpack 2 w/ Code Splitting, Tree Shaking, and other Optimizations

### Tickets / Documentation

- [STENCIL-3250 - Upgrade to Webpack2](https://jira.bigcommerce.com/browse/STENCIL-3250)
- [STENCIL-3245 - Implement Tree Shaking](https://jira.bigcommerce.com/browse/STENCIL-3245)
- [STENCIL-3244 - Adjust webpack.conf to use babel env present](https://jira.bigcommerce.com/browse/STENCIL-3244)

### Depends On

- [STENCIL-286](https://github.com/bigcommerce/stencil-cli/pull/286) - merged
- #961 - merged

#### Stats
[Before](https://rawgit.com/mjschock/stencil/STENCIL-3250-stats/statistics-before.html); size of bundle:  616K; size of zip:  10.3MB
[After](https://rawgit.com/mjschock/stencil/STENCIL-3250-stats/statistics-after-without-vendor-bundle-last3-versions.html); total size of combined bundles:  612K; size of zip: 10.3MB
Note - if we get rid of code splitting (and async/await) the zip remains at 10.3MB but the bundle reduces to 560K (see the alternate diff [here](https://github.com/bigcommerce/cornerstone/compare/master...mjschock:STENCIL-3250-without-code-splitting?expand=1))

Code splitting example:
http://g.recordit.co/REeImIsk0o.gif

#### Appreciation
Thank you @dstaley for the incredibly useful writeup <a href="https://dstaley.com/2017/02/12/stencil-and-webpack-2/">here</a>.

@bigcommerce/stencil-team @bookernath

### CODE REVIEW

File: assets/js/app.js [hunk 1]
```diff
@@ -1,70 +1,50 @@
-import 'babel-polyfill';
+import 'babel-polyfill'; // TODO: can we remove this? see https://jira.bigcommerce.com/browse/MERC-684 and https://github.com/bigcommerce/cornerstone/commit/0648a3f900e093fb3ce8d30ddb087feb8e6b3c6d
```
Comment (mjschock) [inline assets/js/app.js]: want to see if we can actually remove this and just target es3/5 directly.

File: webpack.conf.js [hunk 2]
```diff
@@ -1,28 +1,71 @@
...
+    watch: false,
     devtool: 'source-map',
+    context: __dirname,
+    entry: {
+        main: './assets/js/app.js'
+    },
+    output: {
+        filename: '[chunkhash].theme-bundle.[name].js',
+        path: path.resolve(__dirname, "assets/dist"),
+        publicPath: 'assets/dist/', // TODO: cdnify - https://github.com/bigcommerce/paper/blob/master/index.js#L154
```
Comment (mjschock) [inline webpack.conf.js]: @mcampa - i'm wondering how we can make this work here.  we want the publicPath to include the CDN for the code-split files.  here's some more info:  https://webpack.js.org/guides/public-path/

File: webpack.conf.js [hunk 3]
```diff
@@ -1,28 +1,71 @@
...
-                    cacheDirectory: true,
-                    presets: ['es2015-loose'],
+                use: {
+                    loader: 'babel-loader',
+                    options: {
+                        compact: false,
+                        cacheDirectory: true,
+                        presets: [['es2015', {loose: true, modules: false}]],
+                        plugins: [
+                            'syntax-dynamic-import', // do we need babel-plugin-dynamic-import-webpack?
```
Comment (mjschock) [inline webpack.conf.js]: didn't seem to need this but want to make sure...

File: package.json [hunk 4]
```diff
@@ -37,14 +42,15 @@
...
     "lazysizes": "3.0.0",
     "load-grunt-config": "0.17.1",
     "lodash": "^3.5.0",
     "nod-validate": "^2.0.12",
     "pace": "hubspot/pace#a03f1f1de62c9cea6c88b2267b8d7a83858b6cb6",
     "slick-carousel": "1.5.5",
     "time-grunt": "^1.2.2",
-    "webpack": "^1.14.0"
+    "webpack": "^2.2.1",
+    "webpack-visualizer-plugin": "^0.1.11"
```
Comment (mjschock) [inline package.json]: TODO: remove this before merging

File: stencil.conf.js [hunk 5]
```diff
@@ -27,57 +18,60 @@ var watchOptions = {
...
- * Be sure to call the `done()` callback when finished
- * @param {function} done
  */
-function production(done) {
-    var compiler;
-
+function production() {
+    webpackConfig.watch = false;
     webpackConfig.devtool = false;
-    webpackConfig.plugins.push(new webpack.optimize.DedupePlugin());
```
Comment (mjschock) [inline stencil.conf.js]: https://webpack.js.org/guides/migrating/#dedupeplugin-has-been-removed

File: webpack.conf.js [hunk 6]
```diff
@@ -1,28 +1,71 @@
...
-                query: {
-                    compact: false,
-                    cacheDirectory: true,
-                    presets: ['es2015-loose'],
+                use: {
+                    loader: 'babel-loader',
+                    options: {
+                        compact: false,
+                        cacheDirectory: true,
+                        presets: [['es2015', {loose: true, modules: false}]],
```
Comment (mjschock) [inline webpack.conf.js]: modules: false needed for tree-shaking; not sure if we need loose, not sure what it was here for originally.

Comment (mjschock) [inline webpack.conf.js]: loose mode was for ie10:  https://github.com/bigcommerce/cornerstone/commit/f6f2ec08cd026b478654eedd721565e78d0f9ced

File: assets/js/app.js [hunk 7]
```diff
@@ -1,70 +1,51 @@
-import 'babel-polyfill';
+__webpack_public_path__ = window.__webpack_public_path__; // eslint-disable-line
```
Comment (mjschock) [inline assets/js/app.js]: this sets the public path for webpack to the base path for the cdn

File: templates/layout/amp-iframe.html [hunk 8]
```diff
@@ -23,8 +23,11 @@
         <div class="icons-svg-sprite">{{> components/common/icons/icon-defs }}</div>
 
         {{#block "page"}} {{/block}}
-
-        <script src="{{cdn 'assets/dist/theme-bundle.js'}}"></script>
+        
+        <script>window.__webpack_public_path__ = "{{cdn 'assets/dist/'}}";</script>
```
Comment (mjschock) [inline templates/layout/amp-iframe.html]: the cdn base path is needed for the dynamically loaded bundles

File: webpack.conf.js [hunk 9]
```diff
@@ -1,28 +1,58 @@
...
+        main: './assets/js/app.js',
+    },
+    output: {
+        filename: 'theme-bundle.[name].js',
+        path: path.resolve(__dirname, "assets/dist")
+    },
     bail: true,
     module: {
-        loaders: [
+        rules: [
```
Comment (mjschock) [inline webpack.conf.js]: https://webpack.js.org/guides/migrating/#module-loaders-is-now-module-rules

File: webpack.conf.js [hunk 10]
```diff
@@ -1,28 +1,58 @@
...
             {
                 test: /\.js$/,
-                loader: 'babel',
                 include: /(assets\/js|assets\\js|stencil-utils)/,
-                query: {
-                    compact: false,
-                    cacheDirectory: true,
-                    presets: ['es2015-loose'],
+                use: {
+                    loader: 'babel-loader',
```
Comment (mjschock) [inline webpack.conf.js]: https://webpack.js.org/guides/migrating/#automatic-loader-module-name-extension-removed

File: stencil.conf.js [hunk 11]
```diff
@@ -27,57 +17,61 @@ var watchOptions = {
...
     webpackConfig.plugins.push(new webpack.optimize.UglifyJsPlugin({
-        comments: false
+        comments: false,
+        compress: {
+            warnings: true,
+        },
+        minimize: true,
+        sourceMap: true,
```
Comment (mjschock) [inline stencil.conf.js]: https://webpack.js.org/guides/migrating/#uglifyjsplugin-sourcemap

File: stencil.conf.js [hunk 12]
```diff
@@ -27,57 +17,61 @@ var watchOptions = {
...
-
+function production() {
+    webpackConfig.watch = false;
     webpackConfig.devtool = false;
-    webpackConfig.plugins.push(new webpack.optimize.DedupePlugin());
     webpackConfig.plugins.push(new webpack.optimize.UglifyJsPlugin({
-        comments: false
+        comments: false,
+        compress: {
+            warnings: true,
```
Comment (mjschock) [inline stencil.conf.js]: https://webpack.js.org/guides/migrating/#uglifyjsplugin-warnings

File: webpack.conf.js [hunk 13]
```diff
@@ -1,28 +1,58 @@
...
+            watch: false
+        }),
         new webpack.ProvidePlugin({
             $: 'jquery',
             jQuery: 'jquery',
             'window.jQuery': 'jquery'
+        }),
+        new webpack.optimize.CommonsChunkPlugin({
+            name: 'vendor',
+            minChunks: function (module) {
```
Comment (mjschock) [inline webpack.conf.js]: https://webpack.js.org/plugins/commons-chunk-plugin/#passing-the-minchunks-property-a-function

File: stencil.conf.js [hunk 14]
```diff
@@ -27,57 +17,60 @@ var watchOptions = {
...
+        },
+        sourceMap: true,
```
Comment (mjschock) [inline stencil.conf.js]: https://webpack.js.org/guides/migrating/#uglifyjsplugin-sourcemap

File: webpack.conf.js [hunk 15]
```diff
@@ -1,28 +1,61 @@
...
             }
         ]
     },
     plugins: [
+        new CleanWebpackPlugin(['assets/dist'], {
+            verbose: true,
+            watch: false
+        }),
+        new webpack.LoaderOptionsPlugin({
+             minimize: true
```
Comment (mjschock) [inline webpack.conf.js]: https://webpack.js.org/guides/migrating/#uglifyjsplugin-minimize-loaders

File: assets/js/app.js [hunk 16]
```diff
@@ -1,70 +1,53 @@
...
-import rss from './theme/rss';
-import page from './theme/page';
-import product from './theme/product';
-import search from './theme/search';
-import sitemap from './theme/sitemap';
-import subscribe from './theme/subscribe';
-import wishlist from './theme/wishlist';
 
+const getAccount = () => import('./theme/account');
+const getLogin = () => import('./theme/auth');
```
Comment (mjschock) [inline assets/js/app.js]: these two functions are shared

File: browserlist [hunk 17]
```diff
@@ -0,0 +1,8 @@
+# See https://support.bigcommerce.com/articles/Public/Themes-Supported-Browsers/
```
Comment (mjschock) [inline browserlist]: perhaps these should go back 3 versions to match the autoprefixer settings?

File: stencil.conf.js [hunk 18]
```diff
@@ -41,8 +41,16 @@ function development() {
 function production() {
     webpackConfig.watch = false;
     webpackConfig.devtool = false;
-    webpackConfig.plugins.push(new webpack.optimize.DedupePlugin());
-    webpackConfig.plugins.push(new webpack.optimize.UglifyJsPlugin({ comments: false }));
+    webpackConfig.plugins.push(new webpack.LoaderOptionsPlugin({
+        minimize: true,
```
Comment (mjschock) [inline stencil.conf.js]: https://webpack.js.org/guides/migrating/#uglifyjsplugin-minimize-loaders

File: stencil.conf.js [hunk 19]
```diff
@@ -41,8 +41,16 @@ function development() {
...
+    }));
+    webpackConfig.plugins.push(new webpack.optimize.UglifyJsPlugin({
+        comments: false,
+        compress: {
+            warnings: true,
```
Comment (mjschock) [inline stencil.conf.js]: https://webpack.js.org/guides/migrating/#uglifyjsplugin-warnings

File: templates/layout/amp-iframe.html [hunk 20]
```diff
@@ -24,7 +24,10 @@
 
         {{#block "page"}} {{/block}}
 
-        <script src="{{cdn 'assets/dist/theme-bundle.js'}}"></script>
+        <script>window.__webpack_public_path__ = "{{cdn 'assets/dist/'}}";</script>
```
Comment (mjschock) [inline templates/layout/amp-iframe.html]: this is used (in apps.js) to prefix the dynamically loaded chunk requests

File: webpack.conf.js [hunk 21]
```diff
@@ -1,36 +1,80 @@
...
-        })
+        }),
+        new webpack.optimize.CommonsChunkPlugin({
+            name: 'vendor',
+            minChunks: function (module) {
+                // this assumes your vendor imports exist in the node_modules directory
+                return module.context && module.context.indexOf('node_modules') !== -1;
+            },
+        }),
+        new webpack.optimize.CommonsChunkPlugin({
```
Comment (mjschock) [inline webpack.conf.js]: this one pulls duplicate child files out of the code split chunks into their parent

File: stencil.conf.js [hunk 22]
```diff
@@ -41,8 +41,16 @@ function development() {
 function production() {
     webpackConfig.watch = false;
     webpackConfig.devtool = false;
-    webpackConfig.plugins.push(new webpack.optimize.DedupePlugin());
```
Comment (mjschock) [inline stencil.conf.js]: https://webpack.js.org/guides/migrating/#dedupeplugin-has-been-removed

File: webpack.conf.js [hunk 23]
```diff
@@ -1,36 +1,80 @@
...
+            children: true,
+            minChunks: 2,
+        }),
     ],
+    resolve: {
+        alias: {
+            jquery: path.resolve(__dirname, 'node_modules/jquery/dist/jquery.min.js'),
+            jstree: path.resolve(__dirname, 'node_modules/jstree/dist/jstree.min.js'),
```
Comment (mjschock) [inline webpack.conf.js]: both package.json's for jquery and jstree point to the their non-minified versions, this resolves to the minified versions instead

File: webpack.conf.js [hunk 24]
```diff
@@ -1,36 +1,80 @@
...
             $: 'jquery',
             jQuery: 'jquery',
             'window.jQuery': 'jquery',
-        })
+        }),
+        new webpack.optimize.CommonsChunkPlugin({
+            name: 'vendor',
+            minChunks: function (module) {
+                // this assumes your vendor imports exist in the node_modules directory
+                return module.context && module.context.indexOf('node_modules') !== -1;
```
Comment (mjschock) [inline webpack.conf.js]: pull out the node_modules into the vendor bundle (https://webpack.js.org/plugins/commons-chunk-plugin/#passing-the-minchunks-property-a-function)

File: assets/js/app.js [hunk 25]
```diff
@@ -80,11 +63,16 @@ window.stencilBootstrap = function stencilBootstrap(templateFile, contextJSON =
...
-            $(() => {
-                let pageClass;
+            $(async () => {
                 let globalClass;
+                let pageClass;
+                let PageClass;
+
                 // Finds the appropriate class from the pageType.
-                const PageClass = pageClasses[templateFile];
+                const templatePath = pageClasses[templateFile];
```
Comment (mcampa) [inline assets/js/app.js]: Can you rename `templatePath` to something else? since is not a path anymore

File: browserlist (hunk 17 above)
Comment (mcampa) [inline browserlist]: Yes, change it to last 3

File: stencil.conf.js (hunk 14 above)
Comment (mcampa) [inline stencil.conf.js]: If this plugin is only used to bundle for production we don't include the source maps

File: assets/js/app.js (hunk 25 above)
Comment (mjschock) [inline assets/js/app.js]: sure, how about `pageClassImporter`?

Comment (mcampa) [inline assets/js/app.js]: sure

File: browserlist (hunk 17 above)
Comment (mjschock) [inline browserlist]: done

File: stencil.conf.js (hunk 14 above)
Comment (mjschock) [inline stencil.conf.js]: it's off now, with a comment

**Comment by mjschock** [comment]
[Cornerstone-1.6.2.zip](https://github.com/bigcommerce/cornerstone/files/862293/Cornerstone-1.6.2.zip)

**Comment by nickdengler** [comment]
LGTM 💚 

Ping me when it goes to integration so I can submit a PR for failing tests (https://github.com/bigcommerce/bigcommerce/pull/18253)

**Comment by mjschock** [comment]
[Cornerstone-1.6.3.zip](https://github.com/bigcommerce/cornerstone/files/883425/Cornerstone-1.6.3.zip)

**Pull request merged**